Documentation/design-docs/host-testing.md

```diff
@@ -0,0 +1,100 @@
+
+### End-to-End tests
+This is almost all tests in the `HostActivation` test project currently.
+The tests prepare folders on the disk which emulate the product installation by having
+* The muxer `dotnet` or `apphost`
+* The `hostfxr` in the right place
+* Shared frameworks (or for self contained apps in the app)
+
+For the most part all these are real components. The shared frameworks contain a real `Microsoft.NETCore.App` copy, only higher level frameworks are "mocked" by creating only the necessary files.
+Almost all tests then execute a test application using this test product installation and observe the behavior of the app. This is done by
+* Having the app output information to standard output
+* Turning on host tracing and looking for certain text in the tracing output
+* Exit code of the app
+
+Pros:
+* True End-to-End tests which test real product binaries in the real-world conditions (mostly)
+* Tests are written in C# - so natural integration with test infra, VS UI and so on
+Cons:
+* All tests are out-of-proc which makes debugging product complicated
+* Tests perform large setup work and copy/write lot of files on disk - tests are slow for this reason.
+* Tests run the entire product, not just the host. This includes starting the runtime, JITing and so on. Vast majority of tests don't actually use that and effectively ignore the part of running an actual app code - makes the tests slow.
+
+Going forward we would keep these tests and add new ones to provide true End-to-End coverage.
+
+### Native API tests
+Small portion of the tests actually call specific exports on `hostfxr` or `hostpolicy` directly to test these. All these tests are in the `HostActivation` project under the `GivenThatICareAboutNativeHostApi` test class.
+These tests use a special test application (managed) which invokes the selected exports through PInvokes and performs the testing. The `HostActivation` test only prepares this app and executes it, looking for pieces of its output to verify the outcome.
+Ideally we would migrate these over time to the proposed Component tests infra.
+
+Pros:
+* Testing real product binaries
+* Tests are written in C#, but in a separate project - no direct test infra integration or VS UI support. This is worked around by effectively having "wrappers" for these tests.
+* Direct calls to exports in a controlled environment without running the whole product (mostly)
+Cons:
+* Still runs out-of-proc making debugging harder
+
+## Test proposal
+
+### Unit tests
+Add the ability to write true unit tests. So tests written in C++ which can directly call methods/classes from the product. These tests would be compiled as native executables. For simplicity of the build system, they would live in the product source tree (`src` folder) and would be compiled during the product build. They would combine the test source files with product source files into one executable.
```

Comment:
Will they live in separate unit-test sub-directories of the source directories? 
CMake can handle files from different directories, so I don't think we should mix product files and test files in the same directory.

Reply:
Definitely in separate folders... the simplification comes from the fact that they would belong to the `src/corehost` because that already has a native build setup. Enabling native build in some other subtree is just not worth it.